library/ssl_misc.h

```diff
@@ -1161,10 +1161,10 @@ struct mbedtls_ssl_transform {
     unsigned char out_cid[MBEDTLS_SSL_CID_OUT_LEN_MAX];
 #endif /* MBEDTLS_SSL_DTLS_CONNECTION_ID */
 
-#if defined(MBEDTLS_SSL_CONTEXT_SERIALIZATION)
+#if defined(MBEDTLS_SSL_CONTEXT_SERIALIZATION) || defined(MBEDTLS_SSL_PROTO_TLS1_2)
```

Comment:
64 extra bytes is quite a lot to add into an LTS (assuming you'd like to get this into 3.6).

Would you be happy to instead only enable the TLS-Exporter when `MBEDTLS_SSL_CONTEXT_SERIALIZATION` is enabled? It is enabled in the default configuration so that shouldn't be too disruptive.

Reply:
It wouldn't bother me for my use case, but it would be very unintuitive.

Also: In TLS 1.3 we don't need `randbytes` for the exporter, so I wouldn't like to gate it behind `MBEDTLS_SSL_CONTEXT_SERIALIZATION` for TLS 1.3. But then what to do if we build with both TLS 1.2 and 1.3 enabled and serialization disabled?

That said, if we can't find a better solution then I'm ok with it.

Reply:
I made it so that `mbedtls_ssl_export_keying_material` exists if either `MBEDTLS_SSL_CONTEXT_SERIALIZATION` is defined or `MBEDTLS_SSL_PROTO_TLS1_2` is *not* defined. But maybe that's too convoluted?

What do you think about making a `MBEDTLS_SSL_ENABLE_EXPORT_KEYING_MATERIAL` define in mbedls_config.h instead?

Reply:
I'm leaning towards the same conclusion that maybe a new config option is best. We try to avoid adding them if possible as it makes testing more difficult, but this should be fairly self-contained.

I'll have a think about it and get back to you tomorrow.

Reply:
Having not come up with anything better, I agree. Could you add a config option in `mbedtls_config.h`?

Ideally, it would be off by default, but that won't work if you're consuming Mbed TLS binaries built in the default config, for example if you consume a distro package. If you need it to be on in the default config that is probably okay.

In any case it will be turned on in the `full` config.

I think `MBEDTLS_SSL_KEYING_MATERIAL_EXPORT` would be a good name (the 'enable' seems superfluous since that's true of nearly all config options and the order swap seems clearer to me).

Reply:
> Ideally, it would be off by default, but that won't work if you're consuming Mbed TLS binaries built in the default config, for example if you consume a distro package. If you need it to be on in the default config that is probably okay.

Off by default is fine by me. We're compiling and statically linking Mbed TLS. I don't think anyone besides me has asked for this feature on the mailing list, and if someone does, you could always change the default.

Reply:
I've added the option. Let me know what you think!

Reply:
I enabled it by default.

I could see a case for leaving it off by default in 3.6, someone might compile a custom version with serialization off, and then a patch release suddenly makes the SSL context larger, and they need to figure out where that comes from.

Reply:
Usually, if a user is are compiling a custom configuration, they have their own separate config file which they leave unchanged when they upgrade. This means that if we add a new option it is automatically disabled in their config file (because their file doesn't set the new option).

The only situation where this could be a problem is if the user's upgrade process looks like:
1. Get the new version of the source tree
2. Disable a few listed things in `mbedtls_config.h`
3. Build

But I have not come across anything like this in the wild, and I don't think we make any guarantees about it working.

Reply:
That's actually kind of how we do it for [OpenVPN-NL](https://openvpn.fox-it.com). We've got a small number of patches we apply on top of mbed TLS, mostly changing the config file. Of course, since we do make some changes outside the config file, we can't assume any guarantees from you anyway!

Anyway, it doesn't bother me, so feel free to hit the resolve button!

Reply:
Well in that case I stand corrected, I have seen it in the wild and will bear this in mind for future!

But as you say, I think this is the best thing to do in this case, I'll mark this as resolved.